File: swing/userifc_py/swing/resources/jvm_ui/hellojavafxdemo.py
#!/usr/bin/env jython

from __future__ import (absolute_import, division, print_function,
  unicode_literals)

import sys, os, logging, inspect
#from future.builtins import (ascii, filter, hex, map, oct, zip, str)

if 'java' not in sys.platform.lower():
  raise Exception('This package can only be used on Jython.')

#VERSION_UIFORMSWING = os.getenv('VERSION_UIFORMSWING', '2013.04')
##HERE = os.path.dirname(__file__)
##sys.path.extend([os.path.join(HERE, 'uiform_swing-{0}.jar'.format(VERSION_UIFORMSWING))])
#sys.path.extend(['{0}/.m2/repository/org/sandbox/uiform_swing/{1}/uiform_swing-{1}.jar'.format(os.getenv('HOME', '.'), VERSION_UIFORMSWING)])

from java.lang import System
from java.io import FileInputStream
from java.net import URL

#from javafx.embed.swing import JFXPanel
#_ = JFXPanel() # ? prevent JavaFX Toolkit not initialized exception
#from javafx.geometry import Pos
#from javafx.scene.layout import StackPane, Pane, VBox
#from javafx.scene.control import (Label, Button, TextArea, DialogPane,
#  TextField)
from javafx.fxml import FXMLLoader
from javafx.scene import Scene
from javafx.application import Application
from javafx.stage import WindowEvent
#from javafx.event import EventHandler

def userifc_version():
  import platform

  return "(Jython {0}) Java {1} JavaFX {2}".format(platform.python_version(),
    System.getProperty("java.version"), System.getProperty("javafx.version"))

class HelloJavafxDemo(object):
  '''Description:: '''

  def __init__(self, stageX):
    '''Construct object'''
    
    super(HelloJavafxDemo, self).__init__()
    
    self.stage, self.widgets = stageX, {}
    
    #self.parent = StackPane(id='stackpane1', alignment=Pos.CENTER)
    #self.widgets.update({'parent': self.parent, 'pane1': Pane(id='pane1'),
    #  'vbox1': VBox(id='vbox1', alignment=Pos.CENTER),
    #  'label1': Label('label1', id='label1', alignment=Pos.CENTER),
    #  'button1': Button('button1', id='button1', alignment=Pos.CENTER),
    #  'textview1': TextArea(id='textview1'),
    #  'entry1': TextField(id='entry1', promptText='Enter name')})
    #self.widgets['dialog1'] = DialogPane(id='dialog1', headerText='dialog1',
    #  content=self.widgets['entry1'])
    ##self.widgets['vbox1'].children.add(self.widgets['label1'])
    #self.widgets['vbox1'].children.addAll(self.widgets['label1'], 
    #  self.widgets['button1'], self.widgets['textview1'])
    #self.widgets['pane1'].children.add(self.widgets['vbox1'])
    #self.widgets['parent'].children.addAll(self.widgets['pane1'], 
    #  self.widgets['dialog1'])
    #self.widgets['dialog1'].visible = False
    #self.widgets['parent'].setPrefSize(202.0, 208.0)
    #self.widgets['pane1'].setPrefSize(198.0, 205.0)
    #self.widgets['vbox1'].setPrefSize(195.0, 205.0)
    #self.widgets['label1'].setPrefSize(174.0, 38.0)
    #self.widgets['button1'].setPrefSize(174.0, 48.0)
    #self.widgets['textview1'].setPrefSize(174.0, 107.0)
    
    uiform = 'jvm_ui/helloForm-javafx.fxml'
    rsrc_path = os.environ.get('RSRC_PATH', 'resources')
    #self.parent = FXMLLoader.load(URL('file:' + rsrc_path + '/' + uiform))
    fxmlloader = FXMLLoader()
    self.parent = fxmlloader.load(FileInputStream(rsrc_path + '/' + uiform))
    self.widgets.update({'parent': self.parent.lookup('#stackpane1'),
      'pane1': self.parent.lookup('#pane1'),
      'vbox1': self.parent.lookup('#vbox1'),
      'label1': self.parent.lookup('#label1'),
      'button1': self.parent.lookup('#button1'),
      'textview1': self.parent.lookup('#textview1'),
      'dialog1': self.parent.lookup('#dialog1'),
      'entry1': self.parent.lookup('#entry1')})
    
    #self.widgets['button1'].setOnAction(self.Button1EventHandler())
    #self.widgets['entry1'].setOnAction(self.Entry1EventHandler())
    self.widgets['button1'].setOnAction(self.onButton1Action)
    self.widgets['entry1'].setOnAction(self.onEntry1Action)
    
    #self.stage.setOnCloseRequest(self.StageCloseHandler())
    self.stage.setOnCloseRequest(self.onStageClose)
    self.stage.setTitle(self.parent.id)
    self.parent.visible = True

#  class StageCloseHandler(EventHandler):
#    '''Window listener for stage'''
#    
#    def handle(self, event):
#      if WindowEvent.WINDOW_CLOSE_REQUEST == event.eventType:
#        #event.target.close()
#        HelloJavafxDemo.stage.close()
    
#  class Button1EventHandler(EventHandler):
#    '''Action listener for button1'''
#    
#    def handle(self, event):
#      HelloJavafxDemo.widgets['pane1'].visible = False
#      HelloJavafxDemo.widgets['dialog1'].visible = True
#      HelloJavafxDemo.widgets['entry1'].text = ""

#  class Entry1EventHandler(EventHandler):
#    '''Action listener for entry1'''
#    
#    def handle(self, event):
#      HelloJavafxDemo.widgets['dialog1'].visible = False
#      HelloJavafxDemo.widgets['pane1'].visible = True
#      HelloJavafxDemo.widgets['textview1'].text = 'Hello, ' + \
#        HelloJavafxDemo.widgets['entry1'].text + '.'
  
  def onStageClose(self, event):
    '''Window listener for stage'''
    
    if WindowEvent.WINDOW_CLOSE_REQUEST == event.eventType:
      #event.target.close()
      self.stage.close()
  
  def onButton1Action(self, event):
    '''Action listener for button1'''
    
    self.widgets['pane1'].visible = False
    self.widgets['dialog1'].visible = True
    self.widgets['entry1'].text = ""

  def onEntry1Action(self, event):
    '''Action listener for entry1'''
    
    self.widgets['dialog1'].visible = False
    self.widgets['pane1'].visible = True
    self.widgets['textview1'].text = 'Hello, ' + \
      self.widgets['entry1'].text + '.'

class App(Application):
  def start(self, stagePrime):
    params = self.getParameters()
    param0 = params.getRaw()[0] if 0 < len(params.getRaw()) else ''
    
    view = HelloJavafxDemo(stagePrime)
    view.widgets['textview1'].text = params.getNamed()['pretext']
    stagePrime.setScene(Scene(view.parent, 200, 160))
    stagePrime.show()

def main(argv=None):
  pretext = '{0} GUI\n'.format(userifc_version())
  
  #App.launch(App, ['--pretext=' + pretext])
  Application.launch(App, ['--pretext=' + pretext])
  
  return 0

if '__main__' == __name__:
  sys.exit(main(sys.argv[1:]))


# jython -J-cp <path>/jfxrt.jar hellojavafxemo.py
# or
# java -cp <path>/jython-standalone.jar:<path>/jfxrt.jar org.python.util.jython hellojavafxdemo.py

File: javafx/userifc_py/javafx/resources/jvm_ui/hellojavafxdemo.rb
#!/usr/bin/env jruby -w

=begin rdoc
Description:: Brief description
Version:: $Id:$
=end
require 'English'
#require 'jruby/core_ext'

#if !(defined? RUBY_ENGINE && 'jruby' == RUBY_ENGINE)
if !('java' == RUBY_PLATFORM)
  raise 'This package can only be used on JRuby.'
end

require 'java'
#VERSION_UIFORMSWING = ENV['VERSION_UIFORMSWING'] || '2013.04'
##HERE = File.join('.', File.dirname(__FILE__))
##require File.join(HERE, 'uiform_swing-%s.jar' % [VERSION_UIFORMSWING])
#require '%s/.m2/repository/org/sandbox/uiform_swing/%s/uiform_swing-%s.jar' % [ENV['HOME'] || '.', VERSION_UIFORMSWING, VERSION_UIFORMSWING]

java_import 'java.lang.System'
java_import 'java.io.FileInputStream'
java_import 'java.net.URL'

java_import 'javafx.embed.swing.JFXPanel'
_ = JFXPanel.new  # ? prevent JavaFX Toolkit not initialized exception
java_import 'javafx.geometry.Pos'
['StackPane', 'Pane', 'VBox'].each { |o| 
  java_import 'javafx.scene.layout.%s' % o }
['Label', 'Button', 'TextArea', 'DialogPane', 'TextField'].each { |o| 
  java_import 'javafx.scene.control.%s' % o }
java_import 'javafx.fxml.FXMLLoader'
java_import 'javafx.scene.Scene'
java_import 'javafx.application.Application'
java_import 'javafx.stage.WindowEvent'
java_import 'javafx.event.EventHandler'

class HelloJavafxDemo
  attr_reader :widgets, :parent, :stage
  
  def initialize(stageX)
    @stage, @widgets = stageX, {}
=begin
    @parent = StackPane.new #(id='stackpane1', alignment=Pos::CENTER)
    @parent.id = 'stackpane1' ; @parent.alignment = Pos::CENTER
    @widgets.merge!({'parent' => @parent, 'pane1' => Pane.new,
      'vbox1' => VBox.new, 'label1' => Label.new('label1'),
      'button1' => Button.new('button1'), 'textview1' => TextArea.new,
      'entry1' => TextField.new('Enter name')})
    @widgets['pane1'].id = 'pane1' ; @widgets['vbox1'].id = 'vbox1'
    @widgets['label1'].id = 'label1' ; @widgets['button1'].id = 'button1'
    @widgets['textview1'].id = 'textview1' ; @widgets['entry1'].id = 'entry1'
    @widgets['vbox1'].alignment = Pos::CENTER
    @widgets['label1'].alignment = Pos::CENTER
    @widgets['button1'].alignment = Pos::CENTER
    @widgets['dialog1'] = DialogPane.new #(id='dialog1', headerText='dialog1',
    #  content=@widgets['entry1'])
    @widgets['dialog1'].id = 'dialog1'
    @widgets['dialog1'].headerText = 'dialog1'
    @widgets['dialog1'].content = @widgets['entry1']
    #@widgets['vbox1'].children.add(@widgets['label1'])
    @widgets['vbox1'].children.addAll(@widgets['label1'], 
      @widgets['button1'], @widgets['textview1'])
    @widgets['pane1'].children.add(@widgets['vbox1'])
    @widgets['parent'].children.addAll(@widgets['pane1'], 
      @widgets['dialog1'])
    @widgets['dialog1'].visible = false
    @widgets['parent'].setPrefSize(202.0, 208.0)
    @widgets['pane1'].setPrefSize(198.0, 205.0)
    @widgets['vbox1'].setPrefSize(195.0, 205.0)
    @widgets['label1'].setPrefSize(174.0, 38.0)
    @widgets['button1'].setPrefSize(174.0, 48.0)
    @widgets['textview1'].setPrefSize(174.0, 107.0)
=end
    uiform = 'jvm_ui/helloForm-javafx.fxml'
    rsrc_path = ENV.fetch('RSRC_PATH', 'resources')
    #@parent = FXMLLoader.load(URL.new('file:' + rsrc_path + '/' + uiform))
    fxmlloader = FXMLLoader.new
    @parent = fxmlloader.load(FileInputStream.new(rsrc_path + '/' + uiform))
    @widgets.merge!({'parent' => @parent.lookup('#stackpane1'),
      'pane1' => @parent.lookup('#pane1'),
      'vbox1' => @parent.lookup('#vbox1'),
      'label1' => @parent.lookup('#label1'),
      'button1' => @parent.lookup('#button1'),
      'textview1' => @parent.lookup('#textview1'),
      'dialog1' => @parent.lookup('#dialog1'),
      'entry1' => @parent.lookup('#entry1')})
    
    #@widgets['button1'].setOnAction(Button1EventHandler.new)
    #@widgets['entry1'].setOnAction(Entry1EventHandler.new)
    @widgets['button1'].setOnAction { |event| onButton1Action(event) }
    @widgets['entry1'].setOnAction { |event| onEntry1Action(event) }
    
    #@stage.setOnCloseRequest(StageCloseHandler.new)
    @stage.setOnCloseRequest { |event| onStageClose(event) }
    @stage.setTitle(@parent.id)
    @parent.setVisible(true)
  end
  
  private
  
#  class StageCloseHandler
#    include EventHandler
#    # Window listener for stage
#    
#    def handle(event)
#      if WindowEvent::WINDOW_CLOSE_REQUEST == event.eventType
#        #event.target.close()
#        HelloJavafxDemo.stage.close()
#      end
#    end
#  end
  
#  class Button1EventHandler
#    include EventHandler
#    # Action listener for button1
#    def handle(event)
#      HelloJavafxDemo.widgets['frame1'].setVisible(false)
#      HelloJavafxDemo.widgets['dialog1'].setVisible(true)
#      HelloJavafxDemo.widgets['entry1'].text = ''
#    end
#  end

#  class Entry1EventHandler
#    include EventHandler
#    # Action listener for entry1
#    def handle(event)
#      HelloJavafxDemo.widgets['dialog1'].setVisible(false)
#      HelloJavafxDemo.widgets['frame1'].setVisible(true)
#      HelloJavafxDemo.widgets['textview1'].text = 'Hello, ' + \
#        HelloJavafxDemo.widgets['entry1'].text + '.'
#    end
#  end
  
  def onStageClose(event)
    # Window listener for stage
    
    if WindowEvent::WINDOW_CLOSE_REQUEST == event.eventType
      #event.target.close()
      @stage.close()
    end
  end
  
  def onButton1Action(event)
    # Action listener for button1
    
    @widgets['pane1'].setVisible(false)
    @widgets['dialog1'].setVisible(true)
    @widgets['entry1'].text = ""
  end
  
  def onEntry1Action(event)
    # Action listener for entry1
    
    @widgets['dialog1'].setVisible(false)
    @widgets['pane1'].setVisible(true)
    @widgets['textview1'].text = 'Hello, ' + @widgets['entry1'].text + '.'
  end
  
end # end class HelloJavafxDemo

class App < Application
  def start(stagePrime)
    params = getParameters()
    param0 = params.getRaw().length > 0 ? params.getRaw()[0] : ''
    
    view = HelloJavafxDemo.new(stagePrime)
    view.widgets['textview1'].text = params.getNamed()['pretext']
    stagePrime.setScene(Scene.new(view.parent, 200, 160))
    stagePrime.show()
  end
end # end class App

if $PROGRAM_NAME == __FILE__
  pretext = '(JRuby %s) Java %s JavaFX %s GUI' % [JRUBY_VERSION, 
    System.getProperty('java.version'), System.getProperty("javafx.version")]
  
  params = (['--pretext=' + pretext]).to_java :string
  #App.launch(App.become_java!, params)
  Application.launch(App.become_java!, params)
  
  exit 0
end


# jruby -J-cp <path>/jfxrt.jar hellojavafxdemo.rb
# or
# java -cp <path>/jruby-complete.jar:<path>/jfxrt.jar org.jruby.Main hellojavafxdemo.rb